crates/devices/virtio-vsock/src/packet.rs

```diff
@@ -0,0 +1,934 @@
+            return Err(Error::DescriptorLengthTooSmall);
+        }
+
+        let header = mem
+            .get_slice(chain_head.addr(), PKT_HEADER_SIZE)
+            .map_err(Error::InvalidMemoryAccess)?;
+
+        let header_inner = mem
+            .read_obj(chain_head.addr())
+            .map_err(Error::InvalidMemoryAccess)?;
+
+        let mut pkt = Self {
+            header,
+            header_inner,
+            data: None,
+        };
+
+        // If the `len` field of the header is zero, then the packet doesn't have a `data` element.
+        if pkt.len() == 0 {
+            return Ok(pkt);
+        }
+
+        let data_desc = desc_chain.next().ok_or(Error::DescriptorChainTooShort)?;
+
+        if data_desc.is_write_only() {
+            return Err(Error::UnexpectedWriteOnlyDescriptor);
+        }
+
+        // The data buffer should be large enough to fit the size of the data, as described by
+        // the header descriptor.
+        if data_desc.len() < pkt.len() {
+            return Err(Error::DescriptorLengthTooSmall);
+        }
+
+        let data = {
+            mem.get_slice(data_desc.addr(), data_desc.len() as usize)
+                .map_err(Error::InvalidMemoryAccess)?
+        };
+
+        pkt.data = Some(data);
+        Ok(pkt)
+    }
+
+    /// Create the packet wrapper from an RX chain.
+    ///
+    /// There must be two descriptors in the chain, both writable: a header descriptor and a data
+    /// descriptor.
+    pub fn from_rx_virtq_chain<M, T>(
+        mem: &'a M,
+        desc_chain: &mut DescriptorChain<T>,
+    ) -> Result<Self>
+    where
+        M: GuestMemory,
+        <<M as GuestMemory>::R as GuestMemoryRegion>::B: WithBitmapSlice<'a, S = B>,
+        T: Deref,
+        T::Target: GuestMemory,
+    {
+        let chain_head = desc_chain.next().ok_or(Error::DescriptorChainTooShort)?;
+        // All RX buffers must be device-writable.
+        if !chain_head.is_write_only() {
+            return Err(Error::UnexpectedReadOnlyDescriptor);
+        }
+
+        // The packet header should fit inside the head descriptor.
+        if (chain_head.len() as usize) < PKT_HEADER_SIZE {
+            return Err(Error::DescriptorLengthTooSmall);
+        }
+
+        // All RX descriptor chains should have a header and a data descriptor.
+        if !chain_head.has_next() {
+            return Err(Error::DescriptorChainTooShort);
+        }
+
+        let header = mem
+            .get_slice(chain_head.addr(), PKT_HEADER_SIZE)
+            .map_err(Error::InvalidMemoryAccess)?;
+        let data_desc = desc_chain.next().ok_or(Error::DescriptorChainTooShort)?;
+
+        if !data_desc.is_write_only() {
+            return Err(Error::UnexpectedReadOnlyDescriptor);
+        }
+
+        let data = mem
+            .get_slice(data_desc.addr(), data_desc.len() as usize)
+            .map_err(Error::InvalidMemoryAccess)?;
+
+        Ok(Self {
+            header,
+            header_inner: Default::default(),
```

Comment:
Why are we using the default values for `header_inner` instead of initializing it based on `header` similar to `from_tx_virtq_chain `?

Reply:
On the RX path there is no data in the packet's header and data buffers. The driver is providing empty buffers to be later filled by the device.